application/src/main/kotlin/net/corda/v5/application/uniqueness/model/UniquenessCheckResult.kt

```diff
@@ -0,0 +1,130 @@
+        const val RESULT_REJECTED_REPRESENTATION = 'R'
+
+        /**
+         * Converts an Avro error to a [Failure]. Since there's no common superclass
+         * for avro classes, the parameter can be anything, hence the [Any] type.
+         */
+        fun fromExternalError(avroResponse: UniquenessCheckExternalResponse): Failure {
+            return when (val avroResult = avroResponse.result) {
+                is UniquenessCheckExternalResultInputStateConflict -> {
+                    Failure(
+                        Instant.now(),
+                        UniquenessCheckError.InputStateConflict(avroResult.conflictingStates.map {
+                            // FIXME Consuming tx hash is populated as [null] for now
+                            UniquenessCheckStateDetails(UniquenessCheckStateRef.fromString(it), null)
+                        })
+                    )
+                }
+                is UniquenessCheckExternalResultInputStateUnknown -> {
+                    Failure(
+                        Instant.now(),
+                        UniquenessCheckError.InputStateUnknown(avroResult.unknownStates.map {
+                            UniquenessCheckStateRef.fromString(it)
+                        })
+                    )
+                }
+                is UniquenessCheckExternalResultReferenceStateConflict -> {
+                    Failure(
+                        Instant.now(),
+                        UniquenessCheckError.ReferenceStateConflict(avroResult.conflictingStates.map {
+                            // FIXME Consuming tx hash is populated as [null] for now
+                            UniquenessCheckStateDetails(UniquenessCheckStateRef.fromString(it), null)
+                        })
+                    )
+                }
+                is UniquenessCheckExternalResultReferenceStateUnknown -> {
+                    Failure(
+                        Instant.now(),
+                        UniquenessCheckError.ReferenceStateUnknown(avroResult.unknownStates.map {
+                            UniquenessCheckStateRef.fromString(it)
+                        })
+                    )
+                }
+                is UniquenessCheckExternalResultTimeWindowOutOfBounds -> {
+                    Failure(
+                        Instant.now(),
+                        UniquenessCheckError.TimeWindowOutOfBounds(
+                            avroResult.evaluationTimestamp,
+                            avroResult.timeWindowLowerBound,
+                            avroResult.timeWindowUpperBound
+                        )
+                    )
+                }
+                is UniquenessCheckExternalResultMalformedRequest -> {
+                    // FIXME This should be handled differently
+                    throw CordaRuntimeException("")
+                }
+                else -> {
+                    // TODO should this be handled differently?
+                    throw CordaRuntimeException("")
```

Comment:
Yes, but I don't see a reason to defer this? These are legitimate cases that should be properly handled now.

For the malformed request, it may be you need to extend the formerly internal class - there was no reason to have this before (the uniqueness check server doesn't store these) but as these classes are now dual purpose there may now be a need.

Regardless, it feels like we shouldn't be throwing exceptions here.

Reply:
Good point, I was thinking about how to properly handle this but I think this is the best way.